src/DataStructures/StaticDeque.hpp

```diff
@@ -155,14 +166,39 @@ class StaticDeque
   // move any other entries.
   //
   // The start_position_ variable stores the index of the start of the
-  // portion of the array that is in use, in the range [0, Capacity).
+  // portion of the array that is in use, in the range [0, capacity).
   // Care must be taken to keep this value in range when the start of
   // the used memory crosses the end of the storage.  The size_
   // variable stores the number of existing elements, which must be in
-  // the range [0, Capacity].  (This could not be implemented using a
+  // the range [0, capacity].  (This could not be implemented using a
   // "one past the end" index instead of the size, because that index
-  // is the same for size=0 and size=Capacity.)
-  alignas(T) std::byte data_[sizeof(T) * Capacity];
+  // is the same for size=0 and size=capacity.)
+  template <size_t Cap, typename = void>
+  struct Data {
+    static constexpr bool variable_capacity = false;
+    alignas(T) std::byte data[sizeof(T) * Cap];
+    static constexpr size_t capacity = Cap;
+
+    std::byte* get() { return data; }
+    const std::byte* get() const { return data; }
+  };
+
+  // gcc improperly forbids full specializations
+  // (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85282)
+  template <typename Void>
+  struct Data<std::numeric_limits<size_t>::max(), Void> {
+    static constexpr bool variable_capacity = true;
+    // There doesn't seem to be any alignment guarantee on the
```

Comment:
This is the alignment of `*data` matching the alignment of `T` (or actually of any type with a fundamental alignment because none of the allocation code uses `T` except to get the size).  The standard guarantees that arrays of `char`, `unsigned char`, and `std::byte` (but apparently not `signed char`?) created with `operator new[]` will be properly aligned for any type with fundamental alignment, intended for exactly what we are doing here allocating other objects into a buffer.  I couldn't find any guarantee that `std::vector<std::byte>` also guaranteed that, although I'd be pretty shocked if it didn't do so in practice.

According to cppreference this is usually either 8 or 16 byte alignment.  For stricter alignment we could use `std::aligned_alloc` instead of `operator new[]` (and `free` instead of `delete[]`).  I don't think any other changes would be required.  If such types' requirements are properly reported by `alignof` that would be easy.

Reply:
Okay, that sounds good! The only types where I know stricter alignment _can_ be useful SIMD, but even that's not really the case right now. It used to be alignment matter, but a lot less so nowadays :shrug: